### PR TITLE
Add a sourcerer which walks all BUILD files in the repo

### LINKED ISSUE #12: autogenerate a `sources` filegroup for every package
A recursive filegroup only matches files that are not owned by any other package, which means that to create the `kubernetes-src` tarball for releases, we either need to do this outside bazel, or we need to autogen a `sources` filegroup for each BUILD file.

Any objections to just doing this? It'd be a good chance for me to jump into `gazel`.

### PR DESCRIPTION
**Pull request by ixdy**
All `BUILD` files are run through the buildifier formatter.

Optionally, if `.gazelcfg.json` sets `"AddSourcesRules": true`, then add `package-srcs` and `all-srcs` rules to every `BUILD` file.

I've hopefully explained why we need `package-srcs` and `all-srcs` in the README and comments; basically, we want to make sure there's one rule that's always automanaged and one that can be locally customized as needed.

Fixes #12.

### CODE REVIEW

**Comment by spxtr** [comment]
https://en.wikipedia.org/wiki/Sourcery

File: README [hunk 1]
```diff
@@ -45,6 +45,33 @@ rule after initial creation so you can add and managed other
...
+
+The "package-srcs" rule is a glob matching all files in the
+package recursively, but not any files owned by packages in
+subdirectories.
+
+The "all-srcs" rule includes both the "package-srcs" rule and
+the "all-srcs" rules of all subpackages; i.e. //:all-srcs will
+include all files in your repository.
+
+The "package-srcs" rule defauls to private visibility,
```
Comment (spxtr) [inline README]: defaults

File: gazel/gazel.go [hunk 2]
```diff
@@ -621,21 +627,24 @@ func ReconcileRules(pkgPath string, rules []*bzl.Rule, dryRun bool) (bool, error
...
+
 	return writeFile(path, f, true, dryRun)
 }
 
 func reconcileLoad(f *bzl.File, rules []*bzl.Rule) {
 	usedRuleKindsMap := map[string]bool{}
 	for _, r := range rules {
-		usedRuleKindsMap[r.Kind()] = true
+		// TODO: make less fragile
+		if strings.HasPrefix(r.Kind(), "go_") || strings.HasPrefix(r.Kind(), "cgo_") {
```
Comment (spxtr) [inline gazel/gazel.go:646]: Why is this necessary?

File: gazel/sourcerer.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+		// so pass that up to our parent.
+		return children, nil
+	}
+
+	// Enforce formatting the BUILD file, even if we're not adding srcs rules
+	if !hasRules {
+		v.addRules(pkgPath, nil)
+	}
+
+	if !v.cfg.AddSourcesRules {
```
Comment (mikedanese) [inline gazel/sourcerer.go:68]: Why don't we check this at the top of the function or in main()

File: gazel/gazel.go (hunk 2 above)
Comment (mikedanese) [reply gazel/gazel.go:646]: We don't need to import filegroup

**Comment by mikedanese** [comment]
I'm worried about rebases in the "all-srcs" srcs attribute. This seems like a huge amount of generated code for something that is achieved so easily outside of bazel.

**Comment by spxtr** [comment]
The root BUILD looks like this. IIUC, adding or removing packages *anywhere* in the repo means we need to touch the list in `all-srcs`, which might cause lots of merge conflicts.
```
filegroup(
    name = "package-srcs",
    srcs = glob(["**"], exclude=["bazel-*/**", ".git/**"]),
    tags = ["automanaged"],
    visibility = ["//visibility:private"],
)

filegroup(
    name = "all-srcs",
    srcs = [
        ":package-srcs",
        "//api/openapi-spec:all-srcs",
        "//api/swagger-spec:all-srcs",
        "//build:all-srcs",
        "//cluster/addons:all-srcs",
        "//cmd/clicheck:all-srcs",
        "//cmd/cloud-controller-manager:all-srcs",
        "//cmd/gendocs:all-srcs",
        "//cmd/genkubedocs:all-srcs",
        "//cmd/genman:all-srcs",
        "//cmd/genslateyaml:all-srcs",
        "//cmd/genswaggertypedocs:all-srcs",
        "//cmd/genutils:all-srcs",
        "//cmd/genyaml:all-srcs",
        "//cmd/hyperkube:all-srcs",
        "//cmd/kube-apiserver:all-srcs",
        "//cmd/kube-controller-manager:all-srcs",
        "//cmd/kube-discovery:all-srcs",
        "//cmd/kube-dns:all-srcs",
        "//cmd/kube-proxy:all-srcs",
        "//cmd/kubeadm:all-srcs",
        "//cmd/kubectl:all-srcs",
        "//cmd/kubelet:all-srcs",
        "//cmd/kubemark:all-srcs",
        "//cmd/kubernetes-discovery:all-srcs",
        "//cmd/libs/go2idl/client-gen:all-srcs",
        "//cmd/libs/go2idl/conversion-gen:all-srcs",
        "//cmd/libs/go2idl/deepcopy-gen:all-srcs",
        "//cmd/libs/go2idl/defaulter-gen:all-srcs",
        "//cmd/libs/go2idl/go-to-protobuf:all-srcs",
        "//cmd/libs/go2idl/import-boss:all-srcs",
        "//cmd/libs/go2idl/informer-gen:all-srcs",
        "//cmd/libs/go2idl/lister-gen:all-srcs",
        "//cmd/libs/go2idl/openapi-gen:all-srcs",
        "//cmd/libs/go2idl/set-gen:all-srcs",
        "//cmd/linkcheck:all-srcs",
        "//cmd/mungedocs:all-srcs",
        "//docs:all-srcs",
        "//examples:all-srcs",
        "//federation/apis/core:all-srcs",
        "//federation/apis/federation:all-srcs",
        "//federation/client/cache:all-srcs",
        "//federation/client/clientset_generated/federation_clientset:all-srcs",
        "//federation/client/clientset_generated/federation_internalclientset:all-srcs",
        "//federation/cluster:all-srcs",
        "//federation/cmd/federation-apiserver:all-srcs",
        "//federation/cmd/federation-controller-manager:all-srcs",
        "//federation/cmd/genfeddocs:all-srcs",
        "//federation/cmd/kubefed:all-srcs",
        "//federation/pkg/dnsprovider:all-srcs",
        "//federation/pkg/federation-controller:all-srcs",
        "//federation/pkg/kubefed:all-srcs",
        "//federation/registry/cluster:all-srcs",
        "//hack/boilerplate:all-srcs",
        "//pkg/admission:all-srcs",
        "//pkg/api:all-srcs",
        "//pkg/apimachinery:all-srcs",
        "//pkg/apis/abac:all-srcs",
        "//pkg/apis/apps:all-srcs",
        "//pkg/apis/authentication:all-srcs",
        "//pkg/apis/authorization:all-srcs",
        "//pkg/apis/autoscaling:all-srcs",
        "//pkg/apis/batch:all-srcs",
        "//pkg/apis/certificates:all-srcs",
        "//pkg/apis/componentconfig:all-srcs",
        "//pkg/apis/extensions:all-srcs",
        "//pkg/apis/imagepolicy:all-srcs",
        "//pkg/apis/meta/v1:all-srcs",
        "//pkg/apis/policy:all-srcs",
        "//pkg/apis/rbac:all-srcs",
        "//pkg/apis/storage:all-srcs",
        "//pkg/auth/authenticator:all-srcs",
        "//pkg/auth/authorizer:all-srcs",
        "//pkg/auth/group:all-srcs",
        "//pkg/auth/handlers:all-srcs",
        "//pkg/auth/user:all-srcs",
        "//pkg/capabilities:all-srcs",
        "//pkg/client/cache:all-srcs",
        "//pkg/client/chaosclient:all-srcs",
        "//pkg/client/clientset_generated/clientset:all-srcs",
        "//pkg/client/clientset_generated/internalclientset:all-srcs",
        "//pkg/client/conditions:all-srcs",
        "//pkg/client/informers/informers_generated:all-srcs",
        "//pkg/client/leaderelection:all-srcs",
        "//pkg/client/listers/apps/internalversion:all-srcs",
        "//pkg/client/listers/apps/v1beta1:all-srcs",
        "//pkg/client/listers/authentication/internalversion:all-srcs",
        "//pkg/client/listers/authentication/v1beta1:all-srcs",
        "//pkg/client/listers/authorization/internalversion:all-srcs",
        "//pkg/client/listers/authorization/v1beta1:all-srcs",
        "//pkg/client/listers/autoscaling/internalversion:all-srcs",
        "//pkg/client/listers/autoscaling/v1:all-srcs",
        "//pkg/client/listers/batch/internalversion:all-srcs",
        "//pkg/client/listers/batch/v1:all-srcs",
        "//pkg/client/listers/batch/v2alpha1:all-srcs",
        "//pkg/client/listers/certificates/internalversion:all-srcs",
        "//pkg/client/listers/certificates/v1alpha1:all-srcs",
        "//pkg/client/listers/core/internalversion:all-srcs",
        "//pkg/client/listers/core/v1:all-srcs",
        "//pkg/client/listers/extensions/internalversion:all-srcs",
        "//pkg/client/listers/extensions/v1beta1:all-srcs",
        "//pkg/client/listers/imagepolicy/internalversion:all-srcs",
        "//pkg/client/listers/imagepolicy/v1alpha1:all-srcs",
        "//pkg/client/listers/policy/internalversion:all-srcs",
        "//pkg/client/listers/policy/v1alpha1:all-srcs",
        "//pkg/client/listers/policy/v1beta1:all-srcs",
        "//pkg/client/listers/rbac/internalversion:all-srcs",
        "//pkg/client/listers/rbac/v1alpha1:all-srcs",
        "//pkg/client/listers/storage/internalversion:all-srcs",
        "//pkg/client/listers/storage/v1beta1:all-srcs",
        "//pkg/client/metrics:all-srcs",
        "//pkg/client/record:all-srcs",
        "//pkg/client/restclient:all-srcs",
        "//pkg/client/retry:all-srcs",
        "//pkg/client/testdata:all-srcs",
        "//pkg/client/testing/cache:all-srcs",
        "//pkg/client/testing/core:all-srcs",
        "//pkg/client/transport:all-srcs",
        "//pkg/client/typed/discovery:all-srcs",
        "//pkg/client/typed/dynamic:all-srcs",
        "//pkg/client/unversioned:all-srcs",
        "//pkg/cloudprovider:all-srcs",
        "//pkg/controller:all-srcs",
        "//pkg/conversion:all-srcs",
        "//pkg/credentialprovider:all-srcs",
        "//pkg/dns:all-srcs",
        "//pkg/fieldpath:all-srcs",
        "//pkg/fields:all-srcs",
        "//pkg/generated:all-srcs",
        "//pkg/genericapiserver:all-srcs",
        "//pkg/healthz:all-srcs",
        "//pkg/httplog:all-srcs",
        "//pkg/hyperkube:all-srcs",
        "//pkg/kubeapiserver:all-srcs",
        "//pkg/kubectl:all-srcs",
        "//pkg/kubelet:all-srcs",
        "//pkg/kubemark:all-srcs",
        "//pkg/labels:all-srcs",
        "//pkg/master:all-srcs",
        "//pkg/metrics:all-srcs",
        "//pkg/probe:all-srcs",
        "//pkg/proxy:all-srcs",
        "//pkg/quota:all-srcs",
        "//pkg/registry:all-srcs",
        "//pkg/routes:all-srcs",
        "//pkg/runtime:all-srcs",
        "//pkg/security:all-srcs",
        "//pkg/securitycontext:all-srcs",
        "//pkg/selection:all-srcs",
        "//pkg/serviceaccount:all-srcs",
        "//pkg/ssh:all-srcs",
        "//pkg/storage:all-srcs",
        "//pkg/types:all-srcs",
        "//pkg/util:all-srcs",
        "//pkg/version:all-srcs",
        "//pkg/volume:all-srcs",
        "//pkg/watch:all-srcs",
        "//plugin/cmd/kube-scheduler:all-srcs",
        "//plugin/pkg/admission/admit:all-srcs",
        "//plugin/pkg/admission/alwayspullimages:all-srcs",
        "//plugin/pkg/admission/antiaffinity:all-srcs",
        "//plugin/pkg/admission/deny:all-srcs",
        "//plugin/pkg/admission/exec:all-srcs",
        "//plugin/pkg/admission/gc:all-srcs",
        "//plugin/pkg/admission/imagepolicy:all-srcs",
        "//plugin/pkg/admission/initialresources:all-srcs",
        "//plugin/pkg/admission/limitranger:all-srcs",
        "//plugin/pkg/admission/namespace/autoprovision:all-srcs",
        "//plugin/pkg/admission/namespace/exists:all-srcs",
        "//plugin/pkg/admission/namespace/lifecycle:all-srcs",
        "//plugin/pkg/admission/persistentvolume/label:all-srcs",
        "//plugin/pkg/admission/podnodeselector:all-srcs",
        "//plugin/pkg/admission/resourcequota:all-srcs",
        "//plugin/pkg/admission/security:all-srcs",
        "//plugin/pkg/admission/securitycontext/scdeny:all-srcs",
        "//plugin/pkg/admission/serviceaccount:all-srcs",
        "//plugin/pkg/admission/storageclass/default:all-srcs",
        "//plugin/pkg/auth:all-srcs",
        "//plugin/pkg/client/auth:all-srcs",
        "//plugin/pkg/scheduler:all-srcs",
        "//plugin/pkg/webhook:all-srcs",
        "//test/e2e:all-srcs",
        "//test/e2e_node:all-srcs",
        "//test/fixtures:all-srcs",
        "//test/images:all-srcs",
        "//test/integration:all-srcs",
        "//test/list:all-srcs",
        "//test/soak/cauldron:all-srcs",
        "//test/soak/serve_hostnames:all-srcs",
        "//test/utils:all-srcs",
        "//third_party/forked/etcd237/pkg/fileutil:all-srcs",
        "//third_party/forked/etcd237/wal:all-srcs",
        "//third_party/forked/golang/expansion:all-srcs",
        "//third_party/forked/golang/json:all-srcs",
        "//third_party/forked/golang/netutil:all-srcs",
        "//third_party/forked/golang/reflect:all-srcs",
        "//third_party/forked/golang/template:all-srcs",
        "//vendor:all-srcs",
    ],
    tags = ["automanaged"],
)
```

File: README (hunk 1 above)
Comment (ixdy) [reply README]: fixed

File: gazel/gazel.go (hunk 2 above)
Comment (ixdy) [reply gazel/gazel.go:646]: Yeah, iterating through all rules in the BUILD file will include non-Go rules which we shouldn't import.

I added a comment.

File: gazel/sourcerer.go (hunk 3 above)
Comment (ixdy) [reply gazel/sourcerer.go:68]: see the comment on L63 - this is to enforce buildifier formatting on all BUILD files, not just the ones we're autogenerating.

**Comment by ixdy** [comment]
@spxtr I'm hoping that changes to the root-level BUILD should be infrequent, e.g. only when adding a new package to `//pkg`. If someone creates `//pkg/hyperkube/foo/BUILD`, it should only affect `//pkg/hyperkube/BUILD`.

We could also further clean up some of this list by adding a BUILD file to `//pkg/client` which only defines the srcs targets, further isolating `//pkg/client/` changes from the root-level BUILD.

**Comment by ixdy** [comment]
I'm pretty sure this is all necessary if we want to build the release tarballs inside `bazel`, which I think we do, since it's probably more maintainable than https://github.com/kubernetes/kubernetes/blob/master/build/lib/release.sh. I was willing to figure out some hacky way of getting `kubernetes-src.tar.gz` to work, but the final `kubernetes.tar.gz` tarball also requires fairly large portions of the tree, so the hacks were quickly becoming unmanageable.

**Comment by spxtr** [comment]
LGTM. Would it be possible to add `pkg/BUILD`, `federation/BUILD`, &c manually so that the root-level BUILD file doesn't get so monstrous?

**Comment by ixdy** [comment]
Yeah, I can do that when I update the kubernetes repo.

**Comment by mikedanese** [comment]
Ok, consensus on this was it's pretty gnarly and kinda gross but shortest path to a required feature in kubernetes. Should be add a disclaimer of this somewhere?

**Comment by spxtr** [comment]
Sure.
